### PR TITLE
Add shader rendering support for Lattice

### PR DESCRIPTION
**Pull request by deltakosh**

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15700/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15700/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15700/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15700/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15700/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15700/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15700/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15700/merge#BCU1XR#0

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15700/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15700/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15700/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

File: packages/dev/core/src/Meshes/lattice.material.ts [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+     * @param shaderType "vertex" or "fragment"
+     * @param shaderLanguage The shader language to use.
+     * @returns null if no code to be added, or a list of pointName =\> code.
+     */
+    public override getCustomCode(shaderType: string, shaderLanguage = ShaderLanguage.GLSL) {
+        if (shaderType === "vertex") {
+            // we're adding this specific code at the end of the main() function
+            if (shaderLanguage === ShaderLanguage.WGSL) {
+                return {
+                    CUSTOM_VERTEX_DEFINITIONS: `
```
Comment (sebavan) [inline packages/dev/core/src/Meshes/lattice.material.ts:274]: wondering if we should use this or add new injection points ? or move back the injections points in the new code :-)

**Review by Popov72 — APPROVED** [review]

File: packages/dev/core/src/Materials/Textures/rawTexture3D.ts [hunk 2]
```diff
@@ -6,6 +6,27 @@ import type { Nullable } from "../../types";
...
+    public get width(): number {
+        return this._texture ? this._texture.width : 0;
+    }
+
+    /**
+     * Gets the height of the texture
+     */
+    public get height(): number {
+        return this._texture ? this._texture.height : 0;
+    }
```
[the remark this answers is not in the document]
Comment (deltakosh) [reply packages/dev/core/src/Materials/Textures/rawTexture3D.ts:21]: it was but convoluted and returned an object

[the remark this answers is not in the document]
Comment (deltakosh) [reply packages/dev/core/src/Materials/Textures/rawTexture3D.ts:21]: And I assumed this was an useful addition anyway

**Pull request merged**